### ISSUE TITLE
How do I reference mdx files at runtime?

### ISSUE THREAD

**Issue by gajus** [gajus/gajus-com#1]
I have a function called [`getBlogPostHeads`](https://github.com/gajus/gajus-com/blob/c83a2f8d11c7b00b2bd906bad3d682973d071e63/src/routines/getBlogPostHeads.ts#L12-L14).

It is supposed to read the [`blogPosts/`](https://github.com/gajus/gajus-com/tree/main/src/blogPosts) directory and provide metadata about the available blog posts.

Separately, I had [`getBlogPostBody`](https://github.com/gajus/gajus-com/blob/main/src/routines/getBlogPostBody.ts#L14-L18) function which would load the contents of the blog posts.

These functions are called from `/app/sitemap.xml/route.ts`, `/app/blog/rss.xml/route.ts` and [`/app/blog/[blogPostSlug]/page.tsx`](https://github.com/gajus/gajus-com/blob/main/src/app/blog/%5BblogPostSlug%5D/page.tsx) – all of which using `dynamic` rendering, i.e. cannot be computed at build time.

I could not figure out how to do this using Vercel deployments as the deployments strip the files from output directory resulting in a runtime error:

```
[Error: ENOENT: no such file or directory, scandir '/vercel/path0/src/blogPosts'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'scandir',
  path: '/vercel/path0/src/blogPosts'
}
```

It appears that I cannot use  `generateStaticParams` in `/app/blog/[blogPostSlug]/page.tsx` because parts of the page are dynamically generated at runtime (the `BlogPostReactions`).

The same is true for `/app/sitemap.xml/route.ts` and `/app/blog/rss.xml/route.ts` – the static generation would only work if their entire contents is static, which is not the case.

I thought I can just put the files in `public/`, but then it is not clear how to reference the file path because it changes depending on the build environment.

It is worth pointing out that https://vercel.com/guides/loading-static-file-nextjs-api-route examples shows a similar use case. However, I suspect that their example works because it references a static path file which is extract by https://github.com/vercel/nft

In ideal world, `next.config.js` would simply allow me to indicate which files to retain.

**Comment by gajus** [comment]
Turns out there is a way to include these files. Shoutout to this guy:

https://www.reddit.com/r/nextjs/comments/174dyfp/comment/k4a7k1l/?utm_source=share&utm_medium=web2x&context=3

TLDR is that you can trick Next.js into including the files by dumbing down the expression used to discover the files:

```diff
-    const blogPostPaths = await readdir(
-      resolve(dirname(fileURLToPath(import.meta.url)), '../blogPosts'),
-    );
-
-    const blogPosts = [];
-
-    for (const blogPostSlug of blogPostPaths) {
-      const file = await readFile(
-        resolve(
-          dirname(fileURLToPath(import.meta.url)),
-          `../blogPosts`,
-          blogPostSlug,
-          'blogPost.mdx',
-        ),
-      );
+    const blogPostPaths = await readdir(process.cwd() + '/src/blogPosts');
+
+    const blogPosts = [];
+
+    for (const blogPostSlug of blogPostPaths) {
+      const file = await readFile(
+        resolve(process.cwd() + '/src/blogPosts', blogPostSlug, 'blogPost.mdx'),
+      );
```

**Comment by leerob** [comment]
Here's how I do it: https://github.com/leerob/leerob.io/blob/main/app/db/blog.ts

**Comment by gajus** [comment]
Closing. Originally opened this issue to share feedback with Vercel team. This no longer serves purpose.
